keras_cv/models/segmentation/fcn.py

```diff
@@ -0,0 +1,355 @@
+        model_architecture: str, defines the model architecture based on the
+            implementation details present in the paper. The supported
+            architectures are:
+            1. 'fcn8s', a FCN-8S definition
+            2. 'fcn16s', a FCN-16S definition
+            3. 'fcn32s', a FCN-32S definition
+            Defaults to 'fcn8s'.
+        include_resizing: bool, defines whether to include a `tf.keras.layers.Resizing` layer which defaults to a height and width of value 224. Defaults to `False`
+
+    """
+
+    def __init__(
+        self,
+        classes,
+        backbone="vgg16",
+        model_architecture=None,
+        include_resizing=False,
+        **kwargs
+    ):
+        super(FCN, self).__init__(**kwargs)
+
+        self.resize = include_resizing
+        self.num_classes = classes
+        self.backbone = backbone
+        self.model_architecture = model_architecture
+
+        if self.resize:
+            self.resizing_layer = Resizing(
+                height=224, width=224, interpolation="bilinear"
+            )
+
+    def build(self, input_shape):
+        self.height = input_shape[1]
+        self.width = input_shape[2]
+
+        if isinstance(self.backbone, str):
+
+            if self.backbone == "vgg16":
+                # Assumes `channels_last` format
+                if len(input_shape) == 3:
+                    self.vgg_input_shape = (
+                        input_shape[0],
+                        input_shape[1],
+                        input_shape[2],
+                    )
+                elif len(input_shape) == 4:
+                    self.vgg_input_shape = (
+                        input_shape[1],
+                        input_shape[2],
+                        input_shape[3],
+                    )
+                self.backbone_base_model = VGG16(
+                    include_rescaling=False,
+                    include_top=True,
+                    classes=self.num_classes,
+                    input_shape=self.vgg_input_shape,
+                )
+                self.backbone_name = self.backbone
+            elif self.backbone == "vgg19":
+                self.vgg_input_shape = (input_shape[1], input_shape[2], input_shape[3])
+                self.backbone_base_model = VGG19(
+                    include_rescaling=False,
+                    include_top=True,
+                    classes=self.num_classes,
+                    input_shape=self.vgg_input_shape,
+                )
+                self.backbone_name = self.backbone
+            else:
+                raise ValueError(
+                    "Entered `backbone` argument is not a standard available backbone. Valid options are ['vgg16', 'vgg19']"
+                )
+
+        else:
+            if isinstance(self.backbone, tf.keras.models.Model):
+                layer_list = []
+                # Design choice : Either fully reject, or pick up only Conv2D and MaxPooling2D layers and convert Dense to Conv in `build()`
```

Comment:
Shouldn't we be able to extract multi-level spatial maps from ResNets and other similar architectures to be used as well?

Reply:
So that was something I was having a bit of a concern about. With the skip-connections in ResNets or the Parallel-running connections within Inception networks, how would we pick up the specific layers that we need? Because FCNs are quite basic on that note and do not actually support any kind of the skip-connections although there have been some papers which do propose doing so, at the risk of diverging from the original FCN idea.

Reply:
We can treat an entire portion of ResNet as a layer, and abstract away what's going inside from the FCN. If all it needs is a feature map, and ResNet can produce it (while using other mechanisms under the hood), we should be able to use it without a problem?
I'd have to re-read the paper to see what FCN exactly does with these layers, but from first principles, if it uses the outputs of *some* layers (feature maps) to do something, it should be able to use the outputs of layers (feature maps) to do the same operation.

How we get to that feature map can be abstracted in the layer itself

Reply:
While your solution makes sense, I must wonder, would that abstraction also then allow for more customization, maybe even introducing abstracted-out Attention blocks or other more exotic blocks inside? That would end up going against the premise of the Fully-Convolutional Network, then. (But a really really cool model indeed to tinker around with)
If we try to incorporate your solution, then we can look into doing a check within the abstracted-block for any non-Conv/MaxPool layers. 

About what the FCN does for Resnet-style blocks, it hasn't really been addressed from what I could gather. What it does do for `Dense` layers, is convert them into `1x1 Conv2D` layers with the `number of filters = number of units`.

Reply:
Hmm. To unblock FCNs, we might want to limit the usage to VGG-style nets and look into ResNet-FCNs down the line?